### PR TITLE
1950 use circular buffer to increase efficiency of scheduling

### LINKED ISSUE #1950: Use circular buffer to increase efficiency of scheduling
**What Needs to be Done?**

- [ ] First do this
- [ ] Then do that.
- [ ] Also this other thing.

**Is your feature request related to a problem? Please describe.**

**Describe potential solution outcome**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #1950

### CODE REVIEW

File: src/vt/scheduler/queue.h [hunk 1]
```diff
@@ -56,24 +57,48 @@ struct Queue {
...
+    if (buf_.full()) {
+      impl_.emplace(std::move(elm));
+    } else {
+      buf_.push(std::move(elm));
+    }
+  }
 
-  T const& top() { return impl_.front(); }
+  T pop() {
+    auto t = buf_.pop();
```
Comment (PhilMiller) [inline src/vt/scheduler/queue.h:77]: Can `buf` be empty here? If not, please assert so

File: src/vt/utils/container/circular_buffer.h [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+*/
+
+#if !defined INCLUDED_VT_UTILS_CONTAINER_CIRCULAR_BUFFER_H
+#define INCLUDED_VT_UTILS_CONTAINER_CIRCULAR_BUFFER_H
+
+#include "vt/config.h"
+
+namespace vt { namespace util { namespace container {
+
+template <typename T, int size>
```
Comment (PhilMiller) [inline src/vt/utils/container/circular_buffer.h:51]: Could you rename size to capacity, and then rename len() to size(), to be more similar to standard containers

File: src/vt/utils/container/circular_buffer.h [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+  bool empty() const { return head_ == tail_; }
+  bool full() const { return numFree() == 0; }
+  int len() const { return size - 1 - numFree(); }
+
+  template <typename SerializerT>
+  void serialize(SerializerT& s) {
+    s | head_;
+    s | tail_;
+    s | elms_; // this is inefficient, but it's use is footprinting
```
Comment (PhilMiller) [inline src/vt/utils/container/circular_buffer.h:112]: Its

File: src/vt/scheduler/queue.h [hunk 4]
```diff
@@ -56,24 +57,48 @@ struct Queue {
...
+    if (not impl_.empty()) {
```
Comment (PhilMiller) [inline src/vt/scheduler/queue.h:78]: I was thinking of a slightly stronger phrasing of the invariant this maintains, that can make this condition cheaper. If buf is less than full, then impl should be empty. Thus, we should be able to check buf, which is already hot in cache, before having to (much less often) look at impl at all

Comment (PhilMiller) [reply src/vt/scheduler/queue.h:78]: Note that testing the condition will of course have at least one slot free in buf at this point

Comment (lifflander) [reply src/vt/scheduler/queue.h]: Done

File: src/vt/utils/container/circular_buffer.h (hunk 2 above)
Comment (lifflander) [reply src/vt/utils/container/circular_buffer.h]: Done

File: src/vt/utils/container/circular_buffer.h (hunk 3 above)
Comment (lifflander) [reply src/vt/utils/container/circular_buffer.h]: Fixed

File: src/vt/utils/container/circular_buffer.h [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    } else if (head_ < tail_) {
+      return tail_ - head_ - 1;
+    } else {
+      return capacity + tail_ - head_ - 1;
+    }
+  }
+
+  bool empty() const { return head_ == tail_; }
+  bool full() const { return numFree() == 0; }
+  int size() const { return capacity - 1 - numFree(); }
```
Comment (stmcgovern) [inline src/vt/utils/container/circular_buffer.h:106]: size should also be changed to len, as in the tests

Comment (lifflander) [reply src/vt/utils/container/circular_buffer.h:106]: Fixed

**Review by PhilMiller — APPROVED** [review]
Ideally, the unit testing would cover behavior near full as well as near empty. I don't see that as strictly necessary to move this along, though

**Pull request merged**